File: src/invoice.js
const fs = require('fs');

const formatCurrency = (amount) => {
  return `${(Math.round(amount * 100) / 100).toFixed(2)}`;
}

const getData = (configPath) => {
  const data = JSON.parse(fs.readFileSync(configPath));

  const totals = data.items.map(item => item.hours * item.rate);
  const subTotal = formatCurrency(totals.reduce((a, b) => a + b));
  const gst = data.gstIncluded ? 'Included' : formatCurrency(subTotal * 0.1);
  const total = data.gstIncluded ? formatCurrency(subTotal) : formatCurrency(subTotal * 1.1);
  const items = data.items.map(item => ({ ...item, total: formatCurrency(item.hours * item.rate) }));

  return {
    ...data,
    subTotal,
    gst,
    total,
    items,
  };
}

module.exports = { getData };

File: lib/invoice.rb
# frozen_string_literal: true

require "json"

module Invoice
  def format_currency(amount)
    ((amount * 100).round / 100).to_s
  end

  def create_model(config_path)
    raw = File.read config_path
    data = JSON.parse raw

    items = data["items"].map do |item|
      { **item, total: item["hours"] * item["rate"] }
    end

    totals = data["items"].map do |item|
      item["hours"] * item["rate"]
    end

    sub_total = totals.sum
    gst = sub_total * 0.1
    total = data["gst_included"] ? sub_total : sub_total * 1.1

    {
      me: {
        name: data["me"]["name"],
        addr_line1: data["me"]["addrLine1"],
        addr_line2: data["me"]["addrLine2"],
        country: data["me"]["country"],
        abn: data["me"]["abn"]
      },
      client: {
        name: data["client"]["name"],
        phone: data["client"]["phone"],
        abn: data["client"]["abn"]
      },
      invoice_no: data["invoiceNo"],
      date: data["date"],
      items: items.map do |item|
               {
                 name: item["name"],
                 hours: item["hours"],
                 rate: item["rate"],
                 total: (format_currency item[:total])
               }
             end,
      payment: {
        name: data["payment"]["name"],
        bsb: data["payment"]["bsb"],
        account_no: data["payment"]["accountNo"]
      },
      sub_total: (format_currency sub_total),
      gst: data["gst_included"] == true ? "Included" : (format_currency gst),
      total: (format_currency total)
    }
  end
end
